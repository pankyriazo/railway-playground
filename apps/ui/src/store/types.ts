import { Deployment, Edges, Projects } from "@monorepo/types";

export type State = {
  token: string;
  projects: Projects;
  deployments: Record<string, Edges<Deployment>>;
};

interface ActionBase {
  type: string;
  payload?: unknown;
}

interface ResetAction extends ActionBase {
  type: "RESET";
}
interface SetTokenAction extends ActionBase {
  type: "SET_TOKEN";
  payload: string;
}
interface SetProjectsAction extends ActionBase {
  type: "SET_PROJECTS";
  payload: Projects;
}
interface AddDeploymentAction extends ActionBase {
  type: "ADD_DEPLOYMENT";
  payload: Record<string, Edges<Deployment>>;
}
interface RemoveDeploymentAction extends ActionBase {
  type: "REMOVE_DEPLOYMENT";
  payload: string;
}

export type Action =
  | ResetAction
  | SetTokenAction
  | SetProjectsAction
  | AddDeploymentAction
  | RemoveDeploymentAction;

export type Listener = (state: State) => void;

export type Reducer = (state: State, action: Action) => State;
