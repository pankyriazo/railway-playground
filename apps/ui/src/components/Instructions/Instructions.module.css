.container {
  composes: container from global;

  background-color: var(--color-bg--inverse);
  border-radius: var(--border-radius);
  color: var(--color-text--inverse);
  font-size: var(--font-size-xs);
  line-height: var(--line-height);
  margin: var(--spacing-xxl) auto;
  max-width: 15rem;
  position: relative;
  padding: var(--spacing-md);
  text-align: center;
}
