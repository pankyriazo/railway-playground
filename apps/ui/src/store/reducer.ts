import { Action, State } from "./types";
import { initialState } from "./store";

const reducer = (state: State, action: Action): State => {
  switch (action.type) {
    case "RESET":
      return initialState;
    case "SET_TOKEN":
      return { ...state, token: action.payload };
    case "SET_PROJECTS":
      return { ...state, projects: action.payload };
    case "ADD_DEPLOYMENT":
      return {
        ...state,
        deployments: { ...state.deployments, ...action.payload },
      };
    case "REMOVE_DEPLOYMENT":
      const { [action.payload]: _, ...deployments } = state.deployments;
      return { ...state, deployments };
    default:
      return state;
  }
};

export default reducer;
