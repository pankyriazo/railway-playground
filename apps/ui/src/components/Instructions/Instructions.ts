import { Component } from "@/types";
import store from "@/store";

import styles from "./Instructions.module.css";

class Instructions extends Component<HTMLDivElement> {
  private _token: string = "";

  constructor() {
    super();

    this._element = document.createElement("div");

    this._subscribe();

    this.render();
  }

  private _subscribe() {
    return store.subscribe((state) => {
      if (state.token !== this._token) {
        this._token = state.token;

        this.render();
      }
    });
  }

  override render() {
    if (!this._element) {
      return;
    }

    this._element.innerHTML = "";
    this._element.className = "";
    this._element.dataset.testid = "instructions";

    if (this._token.length > 0) {
      return;
    }

    this._element.className = styles.container!;

    const p = document.createElement("p");
    p.textContent =
      "To get started, please enter your Railway access token. ☝️";
    this._element.appendChild(p);
  }
}

export default Instructions;
