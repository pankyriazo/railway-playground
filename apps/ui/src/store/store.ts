import reducer from "./reducer";
import { Action, Listener, Reducer, State } from "./types";

export const initialState: State = {
  token: "",
  projects: {
    data: null,
    errors: [],
  },
  deployments: {},
};

class Store {
  private _state: State;
  private _reducer: Reducer;
  private _listeners: Listener[] = [];

  constructor() {
    this._state = initialState;
    this._reducer = reducer;
  }

  get state(): State {
    return this._state;
  }

  private _notify(): void {
    this._listeners.forEach((listener) => listener(this._state));
  }

  dispatch(action: Action): void {
    this._state = this._reducer(this._state, action);
    this._notify();
  }

  subscribe(listener: Listener): () => void {
    this._listeners.push(listener);

    listener(this._state);

    return () => {
      this._listeners = this._listeners.filter((l) => l !== listener);
    };
  }
}

const store = new Store();

export default store;
