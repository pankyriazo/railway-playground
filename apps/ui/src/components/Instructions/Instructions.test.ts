import { render } from "@/test-helpers";
import store from "@/store";
import Instructions from "./Instructions";

describe("Instructions", () => {
  beforeEach(() => {
    render(Instructions);
  });

  describe("when the token is empty", () => {
    it("should render the correct text", () => {
      const instructions = document.querySelector("[data-testid=instructions]");

      expect(instructions).not.toBeNull();
      expect(instructions).toMatchSnapshot();
    });
  });

  describe("when the token is not empty", () => {
    it("should not render", () => {
      store.dispatch({
        type: "SET_TOKEN",
        payload: "test-token",
      });

      render(Instructions);

      const instructions = document.querySelector("[data-testid=instructions]");

      expect(instructions!.textContent).toBe("");
      expect(instructions).toMatchSnapshot();
    });
  });
});
