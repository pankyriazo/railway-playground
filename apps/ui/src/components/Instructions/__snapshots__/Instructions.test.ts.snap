// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Instructions when the token is empty should render the correct text 1`] = `
<div
  class="container"
  data-testid="instructions"
>
  <p>
    To get started, please enter your Railway access token. ☝️
  </p>
</div>
`;

exports[`Instructions when the token is not empty should not render 1`] = `
<div
  class=""
  data-testid="instructions"
/>
`;
